File: apps/api/app/api/proposals.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.db import get_db
from app.models.entities import Proposal
from app.proposals.service import (
    ProposalServiceError,
    create_or_replace_greenfield_proposal,
    create_or_replace_proposal,
    load_latest_for_company,
    load_proposal_for_audit,
)
from app.schemas.proposal import ProblemItem, ProposalResponse

router = APIRouter(tags=["proposals"])


@router.post("/audits/{audit_id}/proposal", response_model=ProposalResponse)
def post_proposal(
    audit_id: UUID,
    use_ai: bool | None = Query(default=None),
    db: Session = Depends(get_db),
) -> ProposalResponse:
    try:
        proposal = create_or_replace_proposal(db, audit_id, use_ai=use_ai)
    except ProposalServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc
    return _to_response(proposal)


@router.get("/audits/{audit_id}/proposal", response_model=ProposalResponse)
def get_audit_proposal(audit_id: UUID, db: Session = Depends(get_db)) -> ProposalResponse:
    proposal = load_proposal_for_audit(db, audit_id)
    if proposal is None:
        raise HTTPException(status_code=404, detail="Proposta non trovata.")
    return _to_response(proposal)


@router.get("/companies/{company_id}/proposal", response_model=ProposalResponse)
def get_company_proposal(company_id: UUID, db: Session = Depends(get_db)) -> ProposalResponse:
    proposal = load_latest_for_company(db, company_id)
    if proposal is None:
        raise HTTPException(status_code=404, detail="Proposta non trovata.")
    return _to_response(proposal)


@router.post("/companies/{company_id}/greenfield-proposal", response_model=ProposalResponse)
def post_greenfield_proposal(
    company_id: UUID,
    use_ai: bool | None = Query(default=None),
    db: Session = Depends(get_db),
) -> ProposalResponse:
    try:
        proposal = create_or_replace_greenfield_proposal(db, company_id, use_ai=use_ai)
    except ProposalServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc
    return _to_response(proposal)


def _to_response(proposal: Proposal) -> ProposalResponse:
    return ProposalResponse(
        id=proposal.id,
        audit_id=proposal.audit_id,
        company_id=proposal.company_id,
        kind=proposal.kind,
        source=proposal.source,
        prompt_version=proposal.prompt_version,
        summary=proposal.summary,
        priority_problems=[ProblemItem(**item) for item in proposal.priority_problems],
        recommended_service=proposal.recommended_service,
        strategy=proposal.strategy,
        email_subject=proposal.email_subject,
        email_body=proposal.email_body,
        brief=proposal.brief,
        range_min=proposal.range_min,
        range_max=proposal.range_max,
        currency=proposal.currency,
        range_note=proposal.range_note,
        created_at=proposal.created_at,
        updated_at=proposal.updated_at,
    )

File: apps/desktop/src/api/proposals.ts
import { ApiError, getApiBaseUrl } from "./health";

export type ProposalProblem = {
  code: string;
  severity: string;
  title: string;
  recommendation: string;
};

export type Proposal = {
  id: string;
  audit_id: string | null;
  company_id: string | null;
  kind: string;
  source: string;
  prompt_version: string;
  summary: string;
  priority_problems: ProposalProblem[];
  recommended_service: string;
  strategy: string;
  email_subject: string;
  email_body: string;
  brief: string;
  range_min: number;
  range_max: number;
  currency: string;
  range_note: string;
  created_at: string;
  updated_at: string;
};

async function parseJson<T>(response: Response): Promise<T> {
  let payload: T | { detail?: string };
  try {
    payload = (await response.json()) as T;
  } catch {
    throw new ApiError("Risposta del backend non valida.", response.status);
  }
  if (!response.ok) {
    const detail =
      payload && typeof payload === "object" && "detail" in payload && typeof payload.detail === "string"
        ? payload.detail
        : `Errore ${response.status}`;
    throw new ApiError(detail, response.status);
  }
  return payload as T;
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  let response: Response;
  try {
    response = await fetch(`${getApiBaseUrl()}${path}`, init);
  } catch {
    throw new ApiError("Impossibile raggiungere il backend.");
  }
  return parseJson<T>(response);
}

function aiQueryString(useAi?: boolean): string {
  return useAi == null ? "" : `?use_ai=${useAi ? "true" : "false"}`;
}

export function generateProposal(auditId: string, useAi?: boolean): Promise<Proposal> {
  return request(`/audits/${auditId}/proposal${aiQueryString(useAi)}`, { method: "POST" });
}

export function generateGreenfieldProposal(companyId: string, useAi?: boolean): Promise<Proposal> {
  return request(`/companies/${companyId}/greenfield-proposal${aiQueryString(useAi)}`, { method: "POST" });
}

export async function fetchAuditProposal(auditId: string): Promise<Proposal | null> {
  try {
    return await request<Proposal>(`/audits/${auditId}/proposal`);
  } catch (err: unknown) {
    if (err instanceof ApiError && err.status === 404) return null;
    throw err;
  }
}

export async function fetchCompanyProposal(companyId: string): Promise<Proposal | null> {
  try {
    return await request<Proposal>(`/companies/${companyId}/proposal`);
  } catch (err: unknown) {
    if (err instanceof ApiError && err.status === 404) return null;
    throw err;
  }
}
